### ISSUE TITLE
Extract Camera Orientation (Roll, Pitch, Yaw) from Final Camera Data Rotation Matrix

### ISSUE THREAD

**Issue by mchildsCO76** [openMVG/openMVG#1524]
I am trying to back out the accurate camera position and orientation for each image from the final SFM XML format data. I have extracted the 3x3 rotation matrix and center from the <extrinsics> section (see sample below).

	<extrinsics size="dynamic">
		<value0>
			<key>0</key>
			<value>
				<rotation size="dynamic">
					<value0 size="dynamic">
						<value0>0.95702295765603485</value0>
						<value1>-0.28890461902160897</value1>
						<value2>-0.025321524979221303</value2>
					</value0>
					<value1 size="dynamic">
						<value0>-0.11896062223385834</value0>
						<value1>-0.31143538235956247</value1>
						<value2>-0.94279179725551598</value2>
					</value1>
					<value2 size="dynamic">
						<value0>0.26449088618897237</value0>
						<value1>0.90528565863076182</value1>
						<value2>-0.33241908398953363</value2>
					</value2>
				</rotation>
				<center size="dynamic">
					<value0>-1270883.005548367</value0>
					<value1>-4730773.2508338131</value1>
					<value2>4073656.1954067857</value2>
				</center>
			</value>

Of course the camera location is fine as the <center>stores that as the ECEF coordinates. What I'm having trouble with is backing out a roll, pitch, yaw/heading from the rotation matrix.

Can someone confirm that the rotation matrix is stored in the ECEF reference frame and is in row-major order (i.e. the first three value0/value1/value2 are in row 0 if their parent is value0)?

I had been treating the matrix as already in surface-relative space and getting nowhere near the expected values using the Eigen::Matrix3d::eulerAngles call, but I think I'm now realizing this matrix must be in ECEF space, so I first need to get the ECEF Euler angles, the convert those to the lat/lon/altitude reference frame. 

I just wanted some confirmation that I'm not on a completely wrong path and that there isn't an easier way to extract the calculated camera orientation in a reference frame that can be injected back in to EXIF data for an image.

**Comment by pmoulon** [comment]
You can ask https://github.com/openMVG/openMVG/issues/1523 to see how they extracted roll, pitch heading.
The rotation matrix is stored in row-major format.

**Comment by dhcpp** [comment]
The rotation matrix is row-major format.  The rotation matrix is stored in the ECEF space. 

your examples would extract like
```
0.95702295765603485 -0.28890461902160897 -0.025321524979221303
-0.11896062223385834  -0.31143538235956247 -0.94279179725551598
0.26449088618897237 0.90528565863076182 -0.33241908398953363
```

If you want to transform rotation matrix to euler angels. There has formulation documentation:[link](https://www.geometrictools.com/Documentation/EulerAngles.pdf)

You could ues ``openMVG_main_ExportCameraFrustums`` to confirm whether the camera pose is on the ECEF space.

**Comment by mchildsCO76** [comment]
Thanks for the confirmation of the coordinate space and row-major order. I can extract the rotation matrices now and extract the Euler angles in ECEF space.

I am still struggling with converting the ECEF-space Euler angles to Euler angles relative to a local tangent plane (i.e. normal airplane style roll, pitch, and yaw like EXIF stores). I have found a lot of online resources with bits and pieces of the puzzle, but not a simple one-step, run this math to convert kind of thing. I think my 40+ year old brain can't handle picturing the matrix math like my college-years brain did, so without a worked example to know if I'm doing things right I'm wandering in the dark a bit.

**Comment by mchildsCO76** [comment]
I am continuing to struggle with this. I have completed working and verified code to take an ECEF roll/pitch/yaw at a location and convert it to a roll/pitch/yaw in the traditional lat/lon space. This is based on https://apps.dtic.mil/dtic/tr/fulltext/u2/a484864.pdf and the sample on page 20 verifies with their inputs and expected outputs, so that part is working.

In the sample output from OpenMVG that I'm working with (see above for raw XML), I have an image with the following rotation matrix calculated by OpenMVG:

Rotation Matrix:
0.95702,-0.28890,-0.02532
-0.11896,-0.31144,-0.94279
0.26449,0.90529,-0.33242

ECEF Position:-1270883.005548367, -4730773.2508338131, 4073656.1954067857
Lat/Lon/Height Position: 39.93658094° N, 105.03701386° W 1662.7 m

From the image's XMP metadata, the following orientation was saved in the JPG during flight. It indicates the camera is pointed almost due north, with almost no roll, and the camera is pointing about 70 degrees down.

XMP Metadata:
drone-dji:GimbalRollDegree=+0.00
drone-dji:GimbalYawDegree=-0.50
drone-dji:GimbalPitchDegree=-70.00
drone-dji:FlightRollDegree=-2.30
drone-dji:FlightYawDegree=+0.20
drone-dji:FlightPitchDegree=-2.60

EXIF Roll:  - 2.30°
EXIF Pitch: -72.60°
EXIF Yaw:   - 0.30°

I have tried everything I can think of to extract the ECEF roll/pitch/yaw from the OpenMVG rotation matrix, including every axis order possible using Eigen::eulerAngles, as well as direct math from https://robotics.stackexchange.com/questions/8516/getting-pitch-yaw-and-roll-from-rotation-matrix-in-dh-parameter, but none of the calculated ECEF orientations result in anything near the expected lat/lon (EXIF-space) orientations once run through the ECEF-to-NED orientation conversion code that is verified correct.

Can anyone verify what the ECEF-space roll/pitch/yaw should be from the OpenMVG rotation matrix just so I know that I'm extracting that correctly? I've also tried starting from the inverse of the OpenMVG rotation matrix (in case it was for the opposite direction), but no luck there either.
